drivers/at86rf2xx: Updated address API

maribu opened the pull request:
### Contribution description

Changed the address getter and setter functions to avoid byte order confusion. In addition, passing the value via pointer reduces the ROM size a bit on ARM and quite significantly on AVR.

```
BOARD=microduino-corerf make -C examples/gnrc_minimal
   text    data     bss     dec     hex filename
  45590     426    3354   49370    c0da Current master
  45400     426    3354   49180    c01c This PR
```

```
USEMODULE=at86rf233 BOARD=nucleo-f446re make -C examples/gnrc_networking
   text    data     bss     dec     hex filename
  89328     196   18756  108280   1a6f8
  89264     196   18756  108216   1a6b8 This PR
```

### Testing procedure

E.g. `ping6` on a board equipped with an at86rf2xx transceiver should still work.

### Issues/PRs references

Split out of https://github.com/RIOT-OS/RIOT/pull/12600, as this turned out surprisingly complex. (Or I turned out surprisingly incapable of getting this right...)

benpicco commented:
Looks like you got it reversed again.

### master
```
2019-10-29 23:28:22,929 # Iface  7  HWaddr: 79:1A  Channel: 26  Page: 0  NID: 0x23
2019-10-29 23:28:22,934 #           Long HWaddr: 8F:05:5D:30:5F:8F:F9:1A 
2019-10-29 23:28:22,940 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2019-10-29 23:28:22,946 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
2019-10-29 23:28:22,948 #           6LO  IPHC  
2019-10-29 23:28:22,951 #           Source address length: 8
2019-10-29 23:28:22,954 #           Link type: wireless
2019-10-29 23:28:22,960 #           inet6 addr: fe80::8d05:5d30:5f8f:f91a  scope: link  VAL
2019-10-29 23:28:22,963 #           inet6 group: ff02::2
2019-10-29 23:28:22,965 #           inet6 group: ff02::1
2019-10-29 23:28:22,969 #           inet6 group: ff02::1:ff8f:f91a
2019-10-29 23:28:22,972 #           inet6 group: ff02::1a
```

### this PR
```
2019-10-29 23:28:53,297 # Iface  7  HWaddr: 1A:F9  Channel: 26  Page: 0  NID: 0x23
2019-10-29 23:28:53,301 #           Long HWaddr: 1A:F9:8F:5F:30:5D:05:8F 
2019-10-29 23:28:53,308 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2019-10-29 23:28:53,314 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
2019-10-29 23:28:53,316 #           6LO  IPHC  
2019-10-29 23:28:53,319 #           Source address length: 8
2019-10-29 23:28:53,322 #           Link type: wireless
2019-10-29 23:28:53,327 #           inet6 addr: fe80::18f9:8f5f:305d:58f  scope: link  VAL
2019-10-29 23:28:53,330 #           inet6 group: ff02::2
2019-10-29 23:28:53,333 #           inet6 group: ff02::1
2019-10-29 23:28:53,336 #           inet6 group: ff02::1:ff5d:58f
2019-10-29 23:28:53,339 #           inet6 group: ff02::1a
```

miri64 commented on drivers/at86rf2xx/at86rf2xx_getset.c:
```diff
@@ -130,43 +132,38 @@ static const uint8_t dbm_to_rx_sens[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+void at86rf2xx_get_addr_short(const at86rf2xx_t *dev, network_uint16_t *addr)
 {
-    return (dev->netdev.short_addr[0] << 8) | dev->netdev.short_addr[1];
+    memcpy(addr, dev->netdev.short_addr, IEEE802154_SHORT_ADDRESS_LEN);
```
```suggestion
    memcpy(addr->u8, dev->netdev.short_addr, sizeof(addr->u8));
```

or alternatively even just

```suggestion
    addr->u8[0] = dev->netdev.short_addr[0];
    addr->u8[1] = dev->netdev.short_addr[1];
```

for more type safety?

maribu replied:
All those three are guaranteed to have the same result: A pointer to a union or struct will always be the same as the pointer to its first member. And modern compilers rarely generate the call to memcpy if the size is known at compile time, but instead generates the most efficient instruction sequence to do the copying.

maribu replied:
I updated to use `sizeof(*addr)` here. But I didn't change the pointer passed to `memcpy()`, as `addr->u8` and `addr` are guaranteed to have the same value and `memcpy()` uses `void *` anyway.

miri64 commented on drivers/at86rf2xx/at86rf2xx_getset.c:
```diff
@@ -130,43 +132,38 @@ static const uint8_t dbm_to_rx_sens[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
 {
-    dev->netdev.short_addr[0] = (uint8_t)(addr);
-    dev->netdev.short_addr[1] = (uint8_t)(addr >> 8);
+    memcpy(dev->netdev.short_addr, addr, IEEE802154_SHORT_ADDRESS_LEN);
```
Here of course then the same.

miri64 commented on drivers/at86rf2xx/at86rf2xx_getset.c:
```diff
@@ -130,43 +132,38 @@ static const uint8_t dbm_to_rx_sens[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-        ap[i] = dev->netdev.long_addr[i];
-    }
-    return addr;
+    memcpy(addr, dev->netdev.long_addr, IEEE802154_LONG_ADDRESS_LEN);
```
```suggestion
    memcpy(addr, dev->netdev.long_addr, sizeof(*addr));
```

miri64 commented on drivers/at86rf2xx/at86rf2xx_getset.c:
```diff
@@ -130,43 +132,38 @@ static const uint8_t dbm_to_rx_sens[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-void at86rf2xx_set_addr_long(at86rf2xx_t *dev, uint64_t addr)
+void at86rf2xx_set_addr_long(at86rf2xx_t *dev, const eui64_t *addr)
 {
+    memcpy(dev->netdev.long_addr, addr, IEEE802154_LONG_ADDRESS_LEN);
```
```suggestion
    memcpy(dev->netdev.long_addr, addr, sizeof(dev->netdev.long_addr));
```

maribu replied:
I also used `sizeof(*addr)` here for consistency with the other calls.

maribu commented:
> Looks like you got it reversed again.

Apparently, it was wrong in master. The most significant byte of long address of your board in master is `0x8F == 0b10001111`. But the least significant bit of the most significant byte should be cleared and the second least significant bit of that byte should be set. With this PR the most signifact byte now is `0x1a == 0b00011010`, which correctly indicates non-globally unique and non-multicast.

So making the API more fool proof in regard to the byte order fixed that bug :-)

maribu commented:
Oh, the short address is now generated from the most significant bytes, not the least significant bytes. Let me change that back to the original behavior.

maribu commented:
OK. I tested with `examples/gnrc_networking` (trimmed down to only contain a shell IPv6/6LoWPAN and ICMP) with two Microduino-CoreRF:

```
2019-10-30 09:23:57,698 # ifconfig
2019-10-30 09:23:57,708 # Iface  5  HWaddr: 3D:84  Channel: 26  Page: 0  NID: 0x23
2019-10-30 09:23:57,716 #           Long HWaddr: 3E:84:22:3C:A0:27:3D:84 
2019-10-30 09:23:57,729 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2019-10-30 09:23:57,741 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  6LO  
2019-10-30 09:23:57,744 #           IPHC  
2019-10-30 09:23:57,750 #           Source address length: 8
2019-10-30 09:23:57,755 #           Link type: wireless
2019-10-30 09:23:57,766 #           inet6 addr: fe80::3c84:223c:a027:3d84  scope: link  VAL
2019-10-30 09:23:57,772 #           inet6 group: ff02::1
2019-10-30 09:23:57,774 #           
> ping6 fe80::3c84:223c:a027:3d85
2019-10-30 09:24:04,422 #  ping6 fe80::3c84:223c:a027:3d85
2019-10-30 09:24:04,449 # 12 bytes from fe80::3c84:223c:a027:3d85: icmp_seq=0 ttl=64 rssi=-48 dBm time=11.600 ms
2019-10-30 09:24:05,449 # 12 bytes from fe80::3c84:223c:a027:3d85: icmp_seq=1 ttl=64 rssi=-48 dBm time=12.240 ms
2019-10-30 09:24:06,451 # 12 bytes from fe80::3c84:223c:a027:3d85: icmp_seq=2 ttl=64 rssi=-48 dBm time=13.196 ms
2019-10-30 09:24:06,451 # 
2019-10-30 09:24:06,459 # --- fe80::3c84:223c:a027:3d85 PING statistics ---
2019-10-30 09:24:06,469 # 3 packets transmitted, 3 packets received, 0% packet loss
2019-10-30 09:24:06,477 # round-trip min/avg/max = 11.600/12.345/13.196 ms
>
```

(I changed the long address of the second board from `3E:84:22:3C:A0:27:3D:84` to `3E:84:22:3C:A0:27:3D:85` and updated the short address and IPv6 address accordingly. Those boards have no propper CPU ID, so the get the same address after boot.)

miri64 commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -469,13 +469,13 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
         case NETOPT_ADDRESS:
-            assert(len <= sizeof(uint16_t));
-            at86rf2xx_set_addr_short(dev, *((const uint16_t *)val));
+            assert(len == sizeof(network_uint16_t));
```
The expected type for `NETOPT_ADDRESS` is still a byte array, so the assert should be

```suggestion
            assert(len >= sizeof(network_uint16_t));
```

maribu commented on drivers/at86rf2xx/at86rf2xx_netdev.c:
```diff
@@ -469,13 +469,13 @@ static int _set(netdev_t *netdev, netopt_t opt, const void *val, size_t len)
         case NETOPT_ADDRESS_LONG:
-            assert(len <= sizeof(uint64_t));
-            at86rf2xx_set_addr_long(dev, *((const uint64_t *)val));
+            assert(len == sizeof(eui64_t));
```
Done

benpicco commented:
The address is still reversed - is this intended?

#### before
```
2019-10-31 18:45:35,280 # Iface  7  HWaddr: 06:5A  Channel: 26  Page: 0  NID: 0x23
2019-10-31 18:45:35,290 #           Long HWaddr: 79:7D:36:2C:C9:59:06:5A 
2019-10-31 18:45:35,291 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2019-10-31 18:45:35,298 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
2019-10-31 18:45:35,300 #           6LO  IPHC  
2019-10-31 18:45:35,303 #           Source address length: 8
2019-10-31 18:45:35,306 #           Link type: wireless
2019-10-31 18:45:35,311 #           inet6 addr: fe80::7b7d:362c:c959:65a  scope: link  VAL
2019-10-31 18:45:35,314 #           inet6 group: ff02::2
2019-10-31 18:45:35,317 #           inet6 group: ff02::1
2019-10-31 18:45:35,321 #           inet6 group: ff02::1:ff59:65a
2019-10-31 18:45:35,323 #           inet6 group: ff02::1a
```

#### after
```
2019-10-31 18:46:42,579 # Iface  7  HWaddr: 7D:79  Channel: 26  Page: 0  NID: 0x23
2019-10-31 18:46:42,583 #           Long HWaddr: 5A:06:59:C9:2C:36:7D:79 
2019-10-31 18:46:42,592 #            TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4 
2019-10-31 18:46:42,599 #           AUTOACK  ACK_REQ  CSMA  L2-PDU:102 MTU:1280  HL:64  RTR  
2019-10-31 18:46:42,602 #           6LO  IPHC  
2019-10-31 18:46:42,605 #           Source address length: 8
2019-10-31 18:46:42,608 #           Link type: wireless
2019-10-31 18:46:42,615 #           inet6 addr: fe80::5806:59c9:2c36:7d79  scope: link  VAL
2019-10-31 18:46:42,618 #           inet6 group: ff02::2
2019-10-31 18:46:42,621 #           inet6 group: ff02::1
2019-10-31 18:46:42,625 #           inet6 group: ff02::1:ff36:7d79
2019-10-31 18:46:42,631 #           inet6 group: ff02::1a
```

maribu commented:
> The address is still reversed - is this intended?

Kind of. The code in master markes the first byte as non globally unique and non multicast and then calls ntohll(), effectively marking the least significant byte instead of (as intended) the most significant byte.

So I could keep the call to ntohll() (which doesn't really serve a purpose, as the numeric value of the address doesn't really matter). But in that case I would need to mark the last byte of the LUID instead of the first byte prior to the call of ntohll(),  as this will become the MSB afterwards.

So the L2 address will change no matter what when we want to fix the issue of having 802.15.4 long addresses incorrectly marked as multicast or/and as globally unique. And without the call to ntohll() the code size will be a tad smaller, so I favoured that.

I could hover split this change out of the PR and keep the L2 address as in master for now, if you prefer.

benpicco commented:
I'm not attached to the value or order of the L2 address - just wanted to make sure things are in order. :)

benpicco commented on drivers/at86rf2xx/at86rf2xx_getset.c:
```diff
@@ -130,43 +132,38 @@ static const uint8_t dbm_to_rx_sens[] = { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+void at86rf2xx_get_addr_short(const at86rf2xx_t *dev, network_uint16_t *addr)
 {
-    return (dev->netdev.short_addr[0] << 8) | dev->netdev.short_addr[1];
+    memcpy(addr, dev->netdev.short_addr, sizeof(*addr));
```
Would you save some bytes of ROM by directly assigning the two bytes here?
Otherwise `memcpy` is more consistent.

maribu replied:
The nice thing about modern versions of GCC or clang is that they know about most standard C library functions. E.g. if you use `printf("Hello %s\n", "world");`, they will generate a call to `puts("Hello world");` instead.

E.g. when compiling this code:

```C
#include <string.h>

extern int a, b;

void foo(void)
{
	memcpy(&a, &b, sizeof(a));
}
```

with `arm-none-eabi-gcc -std=c99 -Os -Wall -Wextra -pedantic -Werror -S test.c`, the following assembly is generated:

```asm
	.cpu arm7tdmi
	.eabi_attribute 20, 1
	.eabi_attribute 21, 1
	.eabi_attribute 23, 3
	.eabi_attribute 24, 1
	.eabi_attribute 25, 1
	.eabi_attribute 26, 1
	.eabi_attribute 30, 4
	.eabi_attribute 34, 0
	.eabi_attribute 18, 4
	.file	"test.c"
	.text
	.align	2
	.global	foo
	.arch armv4t
	.syntax unified
	.arm
	.fpu softvfp
	.type	foo, %function
foo:
	@ Function supports interworking.
	@ args = 0, pretend = 0, frame = 0
	@ frame_needed = 0, uses_anonymous_args = 0
	@ link register save eliminated.
	ldr	r3, .L2
	ldr	r2, [r3]
	ldr	r3, .L2+4
	str	r2, [r3]
	bx	lr
.L3:
	.align	2
.L2:
	.word	b
	.word	a
	.size	foo, .-foo
	.ident	"GCC: (Alpine Linux) 9.2.0"
```

With modern C compilers it is safe to assume that they only generate the calls to `memcpy()`, if the call is indeed better than generated machine code to do the copying.

benpicco commented on drivers/at86rf2xx/at86rf2xx.c:
```diff
@@ -106,8 +106,8 @@ void at86rf2xx_reset(at86rf2xx_t *dev)
-    at86rf2xx_set_addr_long(dev, ntohll(addr_long.uint64.u64));
-    at86rf2xx_set_addr_short(dev, ntohs(addr_long.uint16[0].u16));
+    at86rf2xx_set_addr_long(dev, &addr_long);
+    at86rf2xx_set_addr_short(dev, &addr_long.uint16[IEEE802154_LONG_ADDRESS_LEN / sizeof(uint16_t) - 1]);
```
```suggestion
    at86rf2xx_set_addr_short(dev, &addr_long.uint16[ARRAY_SIZE(addr_long.uint16) - 1]);
```
This is easier to understand.

maribu replied:
Done

maribu commented:
Squashed :-)
